File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.py
class ProductOfNumbers(object):

    def __init__(self):
        self.products = [1]

    def add(self, num):
        """
        :type num: int
        :rtype: None
        """
        if num == 0:
            self.products = [1]
        else:
            self.products.append(self.products[-1] * num)
        

    def getProduct(self, k):
        """
        :type k: int
        :rtype: int
        """
        if len(self.products) - 1 < k:
            return 0
        else:
            return self.products[-1] / self.products[-1-k]
        


# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)

File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.js

var ProductOfNumbers = function() {
    this.products = [1]
};

/** 
 * @param {number} num
 * @return {void}
 */
ProductOfNumbers.prototype.add = function(num) {
    if (num === 0){
        this.products = [1]
    }
    else{
        this.products.push(num * this.products[this.products.length - 1]) 
    }
};

/** 
 * @param {number} k
 * @return {number}
 */
ProductOfNumbers.prototype.getProduct = function(k) {
    let n = this.products.length - 1
    let div = n - k
    if (div < 0) return 0
    return Math.round(this.products[n] / this.products[div])
};

/** 
 * Your ProductOfNumbers object will be instantiated and called as such:
 * var obj = new ProductOfNumbers()
 * obj.add(num)
 * var param_2 = obj.getProduct(k)
 */